dll/win32/iernonce/registry.cpp

```diff
@@ -0,0 +1,209 @@
+            break;
+        case REG_ENUM_TYPE::VALUE:
+            Error = RegQueryInfoKeyW(hKey, NULL, 0, NULL,
+                NULL, NULL, NULL, &dwObjectNum, &dwMaxObjNameLen, NULL, NULL, NULL);
+            break;
+        default:
+            assert(0);
+            return FALSE;
+    }
+    if (Error != ERROR_SUCCESS)
+    {
+        return FALSE;
+    }
+
+    LPWSTR RegObjectName = (LPWSTR)HeapAlloc(GetProcessHeap(),
+        0, sizeof(WCHAR) * (dwMaxObjNameLen + 1));
+    if (!RegObjectName)
+    {
+        return FALSE;
+    }
+
+    for (DWORD i = 0; i < dwObjectNum; i++)
+    {
+        DWORD dwcchName = dwMaxObjNameLen + 1;
+
+        switch (EnumType)
+        {
+            case REG_ENUM_TYPE::SUBKEY:
+                Error = hKey.EnumKey(i, RegObjectName, &dwcchName);
+                break;
+            case REG_ENUM_TYPE::VALUE:
+                Error = hKey.EnumValueName(i, RegObjectName, &dwcchName);
+                break;
+            default:
+                assert(0);
+                return FALSE;
+        }
+
+        if (Error != ERROR_SUCCESS)
+        {
+            // TODO: error handling
+            HeapFree(GetProcessHeap(), 0, (PVOID)RegObjectName);
+            return FALSE;
+        }
+
+        if (!Callback(hKey, RegObjectName, dwcchName + 1, Param)) // Call the callback function.
+        {
+            HeapFree(GetProcessHeap(), 0, (PVOID)RegObjectName);
+            return FALSE;
+        }
+    }
+
+    HeapFree(GetProcessHeap(), 0, (PVOID)RegObjectName);
+    return TRUE;
+}
+
+BOOL EnumStringValueCallback(CRegKeyEx &hKey, LPCWSTR lpObjectName, DWORD cchszName, PVOID Param)
+{
+    DWORD dwType;
+    DWORD cbData;
+
+    RunOnceExSection *pSection = (RunOnceExSection *)Param;
+
+    // Query data size
+    if (hKey.QueryValue(lpObjectName, &dwType, NULL, &cbData) != ERROR_SUCCESS)
+    {
+        return FALSE;
+    }
+
+    if (dwType != REG_SZ)
+    {
+        return TRUE;
+    }
+
+    ATL::CStringW Buffer;
+
+    assert((cbData & 1) == 0);
+    LPWSTR szBuffer = Buffer.GetBuffer((cbData / 2) + 1);
+
+    if (hKey.QueryValue(lpObjectName, &dwType, szBuffer, &cbData) != ERROR_SUCCESS)
+    {
+        Buffer.ReleaseBuffer();
+        return FALSE;
+    }
+    szBuffer[cbData / 2] == L'\0';
+    Buffer.ReleaseBuffer();
+
+    if (lpObjectName[0] == L'\0')
+    {
+        // this is the default value
+        pSection->m_SectionTitle = Buffer;
+    }
+    else
+    {
+        pSection->m_EntryMap.insert({ lpObjectName , Buffer });
+    }
+
+    return TRUE;
+}
+
+RunOnceExSection::RunOnceExSection(CRegKeyEx &hParentKey, LPCWSTR lpSubKeyName) : m_SectionName(lpSubKeyName)
+{
+    m_bSuccess = FALSE;
+    CRegKeyEx hKey;
+
+    if (hKey.Open(hParentKey, lpSubKeyName, KEY_READ) != ERROR_SUCCESS)
+    {
+        return;
+    }
+
+    if (!RegEnumObject(hKey, REG_ENUM_TYPE::VALUE, EnumStringValueCallback, (PVOID)this))
+    {
+        hKey.Close();
```

Comment:
hkey handles this, so you don't need to do that manually

Reply:
despite MSDN says that `~CRegKey` will release `m_hKey` [here](https://docs.microsoft.com/en-us/cpp/atl/reference/cregkey-class?view=msvc-160#dtor)

but I found the code in ReactOS actually does nothing when it's been destructed🤣 [ReactOS code here]( https://github.com/reactos/reactos/blob/master/sdk/lib/atl/atlbase.h#L1044)

It looks like a bug... perhaps I should fix it. However, closing the handle manually always works fine I guess, so I wrote it like this 🤣